### PR TITLE
Add bucket verification in TableWriter Fuzzer

### PR DESCRIPTION
**Pull request by kewang1024**

### CODE REVIEW

**Comment by aditi-pandit** [comment]
@kewang1024, @xiaoxmeng : Ke, Meng : I guess we should do extend Fuzzer in a similar manner for bucketed but not partitioned files. I can take that on after this is merged.

File: velox/exec/fuzzer/PrestoQueryRunner.cpp [hunk 1]
```diff
@@ -541,8 +541,8 @@ std::optional<std::string> PrestoQueryRunner::toSql(
 
   // Returns a CTAS sql with specified table properties from TableWriteNode,
   // example sql:
-  // CREATE TABLE tmp_write WITH (PARTITIONED_BY = ARRAY['p0'])
-  // AS SELECT * FROM tmp
+  // CREATE TABLE tmp_write WITH (PARTITIONED_BY = ARRAY['p0'], BUCKETED_COUNT =
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/PrestoQueryRunner.cpp:544]: Shall we also update the document? Thanks!

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 2]
```diff
@@ -263,9 +294,6 @@ std::vector<RowVectorPtr> WriterFuzzer::generateInputData(
   auto inputType = ROW(std::move(names), std::move(types));
   std::vector<RowVectorPtr> input;
 
-  // For partition keys, limit the distinct value to 4 to avoid exceeding
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:266]: This comment still applies?

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 3]
```diff
@@ -390,47 +428,75 @@ std::string WriterFuzzer::getReferenceOutputDirectoryPath(int32_t layers) {
...
+void WriterFuzzer::comparePartitionAndBucket(
     const std::string& outputDirectoryPath,
-    const std::string& referenceOutputDirectoryPath) {
-  const auto partitions = getPartitionNames(outputDirectoryPath);
-  LOG(INFO) << "Velox written partitions:" << std::endl;
-  for (const std::string& partition : partitions) {
-    LOG(INFO) << partition << std::endl;
+    const std::string& referenceOutputDirectoryPath,
+    int32_t bucketCount) {
+  std::vector<std::string> partitionNames;
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:435]:
```
LOG(INFO) << "Velox output directory:" << outputDirectoryPath << std::endl;
const auto partitionNameAndFileCount = getPartitionNameAndFilecount(outputDirectoryPath);
LOG(INFO) << "Partitions and file count:" << std::endl;
std::vector<std::string> partitionNames;
partitionNames.reserve(partitionNameAndFileCount.size());
for (const auto& entry : partitionNameAndFileCount) {
```

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 4]
```diff
@@ -390,47 +428,75 @@ std::string WriterFuzzer::getReferenceOutputDirectoryPath(int32_t layers) {
...
+    LOG(INFO) << i.first << ":" << i.second << std::endl;
+    partitionNames.emplace_back(i.first);
   }
 
-  const auto referencedPartitions =
-      getPartitionNames(referenceOutputDirectoryPath);
-  LOG(INFO) << "Presto written partitions:" << std::endl;
-  for (const std::string& partition : referencedPartitions) {
-    LOG(INFO) << partition << std::endl;
+  std::vector<std::string> referencePartitionNames;
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:445]: ditto: put the var declaration to the place where it is first used? Thanks!

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 5]
```diff
@@ -390,47 +428,75 @@ std::string WriterFuzzer::getReferenceOutputDirectoryPath(int32_t layers) {
...
+  std::map<std::string, int32_t> partitionNameAndFileCount;
 
   for (std::string directory : directories) {
     // If it's a hidden directory, ignore
     if (directory.find("/.") != std::string::npos) {
       continue;
     }
 
+    // Count non-empty non-hidden files
+    auto files = fileSystem->list(directory);
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:482]:
```
const auto files
```

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 6]
```diff
@@ -390,47 +428,75 @@ std::string WriterFuzzer::getReferenceOutputDirectoryPath(int32_t layers) {
...
+    int32_t fileCount = 0;
+    for (const auto& file : files) {
+      if (file.find("/.") == std::string::npos &&
+          fileSystem->openFileForRead(file)->size() > 0) {
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:486]: Why we need to check file size > 0? Is this to check if this is a directory or regular file? Please comment. Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@kewang1024 LGTM % minors. Thanks!

File: velox/exec/fuzzer/WriterFuzzer.cpp (hunk 2 above)
Comment (kewang1024) [reply velox/exec/fuzzer/WriterFuzzer.cpp:266]: I adjusted kMaxPartitionsPerWritersSession from 100 to 400 to give use room for more bucket (up to 5), just adjusted the comment accordingly.

File: velox/exec/fuzzer/WriterFuzzer.cpp (hunk 6 above)
Comment (kewang1024) [reply velox/exec/fuzzer/WriterFuzzer.cpp:492]: Commented in the code

**Review by xiaoxmeng — APPROVED** [review]

**Pull request closed without merging**